File: Easy_Problems/10. Minimum Waiting Time/Python/Solution2.py
# Greedy Approach + Improved Logic
# Intuition is that all remaining queries are waiting for the current query to finish, therefore need to add n times where n is the number of queries remaining
# Time Complexity: O(N · log(N)), where N is the number of queries
# Space Complexity: O(1)

def get_total_min_waiting_time(queries):
    queries.sort()

    total_waiting_time = 0
    
    for idx, duretion in enumerate(queries):
        queries_left = len(queries) - (idx + 1)
        total_waiting_time += queries_left * duretion

    return total_waiting_time


# Tests for get_total_min_waiting_time
assert(get_total_min_waiting_time([3, 2, 1, 2, 6]) == 17)
assert(get_total_min_waiting_time([10]) == 0)
assert(get_total_min_waiting_time([1, 10]) == 1)

File: Easy_Problems/10. Minimum Waiting Time/Javascript/Solution2.js
// Greedy Approach + Improved Logic
// Intuition is that all remaining queries are waiting for the current query to finish, therefore need to add n times where n is the number of queries remaining
// Time Complexity: O(N · log(N)), where N is the number of queries
// Space Complexity: O(1)


function getTotalMinWaitingTime(queries) {
    queries.sort((a, b) => a - b);
  
    let totalWaitingTime = 0;
  
    for (let i = 0; i < queries.length; i++) {
      const duration = queries[i];
      const queriesLeft = queries.length - (i + 1);
      totalWaitingTime += duration * queriesLeft;
    }
  
    return totalWaitingTime;
}


// Tests for getTotalMinWaitingTime
const testGetTotalMinWaitingTime = () => {
    console.log(getTotalMinWaitingTime([3, 2, 1, 2, 6]) == 17)
    console.log(getTotalMinWaitingTime([10]) == 0)
    console.log(getTotalMinWaitingTime([1, 10]) == 1)
}